backend/hidrive/hidrive.go

```diff
@@ -0,0 +1,1021 @@
+	}
+	f.features = (&fs.Features{
+		CanHaveEmptyDirectories: true,
+	}).Fill(ctx, f)
+	f.srv.SetErrorHandler(errorHandler)
+
+	if ts != nil {
+		transaction := func() error {
+			resolvedRoot := f.resolvePath("")
+			_, err := f.fetchMetadataForPath(ctx, resolvedRoot, api.HiDriveObjectNoMetadataFields)
+			return err
+		}
+		f.tokenRenewer = oauthutil.NewRenew(f.String(), ts, transaction)
+	}
+
+	// Do not allow the root-prefix to be non-existent nor a directory,
+	// but it can be empty.
+	if f.opt.RootPrefix != "" {
+		item, err := f.fetchMetadataForPath(ctx, f.opt.RootPrefix, api.HiDriveObjectNoMetadataFields)
+		if err != nil {
+			return nil, fmt.Errorf("could not access root-prefix: %w", err)
+		}
+		if item.Type != api.HiDriveObjectTypeDirectory {
+			return nil, errors.New("The root-prefix needs to point to a valid directory or be empty")
+		}
+	}
+
+	resolvedRoot := f.resolvePath("")
+	item, err := f.fetchMetadataForPath(ctx, resolvedRoot, api.HiDriveObjectNoMetadataFields)
+	if err != nil {
+		if isHTTPError(err, 404) {
+			// NOTE: NewFs needs to work with paths that do not exist,
+			// in case they will be created later (see mkdir).
+			return f, nil
+		}
+		return nil, fmt.Errorf("could not access root-path: %w", err)
+	}
+	if item.Type != api.HiDriveObjectTypeDirectory {
+		fs.Debugf(f, "The root is not a directory. Setting its parent-directory as the new root.")
+		// NOTE: There is no need to check
+		// if the parent-directory is inside the root-prefix:
+		// If the parent-directory was outside,
+		// then the resolved path would be the root-prefix,
+		// therefore the root-prefix would point to a file,
+		// which has already been checked for.
+		// In case the root-prefix is empty, this needs not be checked,
+		// because top-level files cannot exist.
+		f.root = path.Dir(f.root)
+		return f, fs.ErrorIsFile
+	}
+
+	return f, nil
+}
+
+// newObject constructs an Object by calling the given function metaFiller
+// on an Object with no metadata.
+//
+// metaFiller should set the metadata of the object or
+// return an appropriate error.
+func (f *Fs) newObject(remote string, metaFiller func(*Object) error) (fs.Object, error) {
+	o := &Object{
+		fs:     f,
+		remote: remote,
+	}
+	var err error
+	if metaFiller != nil {
+		err = metaFiller(o)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return o, nil
+}
+
+// newObjectFromHiDriveObject constructs an Object from the given api.HiDriveObject.
+func (f *Fs) newObjectFromHiDriveObject(remote string, info *api.HiDriveObject) (fs.Object, error) {
+	metaFiller := func(o *Object) error {
+		return o.setMetadata(info)
+	}
+	return f.newObject(remote, metaFiller)
+}
+
+// NewObject finds the Object at remote.
+//
+// If remote points to a directory then it returns fs.ErrorIsDir.
+// If it can not be found it returns the error fs.ErrorObjectNotFound.
+func (f *Fs) NewObject(ctx context.Context, remote string) (fs.Object, error) {
+	fs.Debugf(f, "executing NewObject(%s).", remote)
```

Comment:
For release we'll probably want to comment out these sort of debugs as they get too noisy. I usually leave them in the code but commented out.